llvm/lib/Transforms/Vectorize/VectorCombine.cpp

```diff
@@ -729,6 +730,116 @@ bool VectorCombine::foldBitcastShuf(Instruction &I) {
+      if (auto *ConstValue = dyn_cast<Constant>(SplattedVal))
+        return ConstValue->isAllOnesValue();
+    return false;
+  };
+  if (!IsAllTrueMask(VPI.getArgOperand(2)))
+    return false;
+
+  // Check to make sure we support scalarization of the intrinsic
+  Intrinsic::ID IntrID = VPI.getIntrinsicID();
+  std::set<Intrinsic::ID> SupportedIntrs(
+      {Intrinsic::vp_smax, Intrinsic::vp_smin, Intrinsic::vp_umax,
+       Intrinsic::vp_umin, Intrinsic::vp_copysign, Intrinsic::vp_minnum,
+       Intrinsic::vp_maxnum});
+  if (!VPBinOpIntrinsic::isVPBinOp(IntrID) && !SupportedIntrs.count(IntrID))
+    return false;
+
+  // Calculate cost of splatting both operands into vectors and the vector
+  // intrinsic
+  VectorType *VecTy = cast<VectorType>(VPI.getType());
+  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+  InstructionCost SplatCost =
+      TTI.getVectorInstrCost(Instruction::InsertElement, VecTy, CostKind, 0) +
+      TTI.getShuffleCost(TargetTransformInfo::SK_Broadcast, VecTy);
+
+  // Calculate the cost of the VP Intrinsic
+  SmallVector<Type *, 4> Args;
+  for (Value *V : VPI.args())
+    Args.push_back(V->getType());
+  IntrinsicCostAttributes Attrs(IntrID, VecTy, Args);
+  InstructionCost VectorOpCost = TTI.getIntrinsicInstrCost(Attrs, CostKind);
+  InstructionCost OldCost = 2 * SplatCost + VectorOpCost;
+
+  // Determine scalar opcode
+  std::optional<unsigned> FunctionalOpcode =
+      VPI.getFunctionalOpcode();
+  bool ScalarIsIntr = false;
+  std::optional<Intrinsic::ID> ScalarIntrID;
+  if (!FunctionalOpcode) {
+    ScalarIntrID = VPI.getFunctionalIntrinsicID();
+    if (!ScalarIntrID)
+      return false;
+    ScalarIsIntr = true;
+  }
+
+  // Calculate cost of scalarizing
+  InstructionCost ScalarOpCost = 0;
+  if (ScalarIsIntr) {
+    IntrinsicCostAttributes Attrs(*ScalarIntrID, VecTy->getScalarType(), Args);
+    ScalarOpCost = TTI.getIntrinsicInstrCost(Attrs, CostKind);
+  } else {
+    ScalarOpCost =
+        TTI.getArithmeticInstrCost(*FunctionalOpcode, VecTy->getScalarType());
+  }
+
+  // The existing splats may be kept around if other instructions use them.
+  InstructionCost CostToKeepSplats =
+      (SplatCost * !Op0->hasOneUse()) + (SplatCost * !Op1->hasOneUse());
+  InstructionCost NewCost = ScalarOpCost + SplatCost + CostToKeepSplats;
+
+  LLVM_DEBUG(dbgs() << "Found a VP Intrinsic to scalarize: " << VPI
+                    << "\n");
+  LLVM_DEBUG(dbgs() << "Cost of Intrinsic: " << OldCost
+                    << ", Cost of scalarizing:" << NewCost << "\n");
+
+  // We want to scalarize unless the vector variant actually has lower cost.
+  if (OldCost < NewCost || !NewCost.isValid())
+    return false;
+
+  // Scalarize the intrinsic
+  ElementCount EC = cast<VectorType>(Op0->getType())->getElementCount();
+  Value *EVL = VPI.getArgOperand(3);
+  const DataLayout &DL = VPI.getModule()->getDataLayout();
+  bool IsKnownNonZeroVL = isKnownNonZero(EVL, DL, 0, &AC, &VPI, &DT);
```

Comment:
Re @topperc 's comment, IIUC the concern is that if VL=0 then we will be performing an operation that wouldn't have otherwise been executed in the VP version. And for div/rem etc. this has the potential of introducing `poison` where there wasn't before.
Hence why we want to check that VL is non zero for these ops.

But the lang ref states that everything past VL should be poison: https://llvm.org/docs/LangRef.html#id2142
(That particular paragraph says undef, but the individual VP intrinsics say poison)

So any VP node with VL=0 should be a vector with all poison elements right? I just checked and InstCombine doesn't canonicalise this, but if that's what the LangRef states then maybe we don't need to restrict these ops to non-zero VLs?

Reply:
Did a bit more investigation, something in SelectionDAG replaces `t9: nxv2i32 = vp_add t2, t4, t6, Constant:i64<0>` with `undef`, so that does seem to be the behaviour.

Reply:
We should canonicalize when the VL is known zero to poison. But here we can only optimize when the VL is known *not* zero. 

The problem with scalarizing is that if there is a division by zero, an exception will occur but if we keep as a VP, no division by zero is performed. I think Craig’s concern was that if the vp would not have introduced an exception, then we shouldn’t add it here due to scalarizing.